File: baseunits/modules/HentaiFox.pas
unit HentaiFox;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, RegExpr;

implementation

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation;
  var Page: Integer; const WorkPtr: Integer; const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL) then
  begin
    Result := NO_ERROR;
    Page := StrToIntDef(XPathString('//*[@class="pagination"]/a[last()-1]', MangaInfo.FHTTP.Document), 1);
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + '/pag/' + IncStr(AURL) + '/') then
  begin
    Result := NO_ERROR;
    XPathHREFAll('//*[@class="galleries_overview"]//*[contains(@class,"item")]/a', MangaInfo.FHTTP.Document, ALinks, ANames);
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := MaybeFillHost(Module.RootURL, XPathString('//*[@class="cover"]/img/@src'));
          if title = '' then title := XPathString('//*[@class="info"]/h1');
          artists := XPathString('//*[@class="info"]/span[starts-with(.,"Artist")]/substring-after(.,": ")');
          genres := XPathString('//*[@class="info"]/string-join(./span/a,", ")');
          chapterLinks.Add(url);
          chapterName.Add(title);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  s: String;
  i: Integer;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL)) then
    begin
      Result := True;
      XPathStringAll('//*[@class="gallery"]//img/@data-src', Document, PageLinks);
      if PageLinks.Count <> 0 then
      begin
        for i := 0 to PageLinks.Count - 1 do
        begin
          if LeftStr(PageLinks[i], 2) = '//' then
            PageLinks[i] := 'https:' + PageLinks[i];
        end;
        PageLinks.Text := ReplaceRegExpr('(?i)(/\d+)t\.', PageLinks.Text, '$1.', True);
      end;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'HentaiFox';
    RootURL := 'https://hentaifox.com';
    Category := 'H-Sites';
    SortedList := True;
    OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/HentaiFox.lua
function getinfo()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  if http.get(mangainfo.url) then
    local x=TXQuery.Create(http.document)
    if mangainfo.title == '' then
      mangainfo.title = x.xpathstring('//div[@class="info"]/h1')
    end
    mangainfo.coverlink=MaybeFillHost(module.RootURL, x.xpathstring('//div[@class="cover"]//img/@src'))
    if module.website == 'HentaiFox' then
      mangainfo.artists=x.xpathstringall('//*[@class="artists"]/li/a')
      mangainfo.genres=x.xpathstringall('//*[@class="tags"]//li/a')
    else
      mangainfo.artists=x.xpathstringall('//div[@class="tags" and contains(h3, "Artist")]/div/a/span/text()')
      mangainfo.genres=x.xpathstringall('//div[@class="tags" and (contains(h3, "Tags") or contains(h3, "Category"))]/div/a/span/text()')
    end
    mangainfo.chapterlinks.add(mangainfo.url)
    mangainfo.chapternames.add(mangainfo.title)
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  task.pagelinks.clear()
  if http.get(MaybeFillHost(module.rooturl, url)) then
    local x=TXQuery.Create(http.Document)
    if module.website == 'HentaiFox' then
      local galleryId = x.xpathstring('//a[@class="g_button"]/@href'):match('/.-/(%d+)/')
      task.pagenumber = tonumber(x.xpathstring('//span[@class="i_text pages" and contains(., "Pages")]/substring-after(.,": ")'))
      for i = 1, task.pagenumber do
        task.PageContainerLinks.Add(MaybeFillHost(module.RootURL, '/g/' .. galleryId .. '/' .. i))
      end
    else
      local v = x.xpath('//*[@class="gallery"]//img/@data-src')
      for i = 1, v.count do
        local s = v.get(i).toString;
        s = s:gsub('^//', 'https://'):gsub('(/%d+)[tT]%.', '%1.')
        task.pagelinks.add(s)
      end
    end
  else
    return false
  end
  return true
end

function getdirectorypagenumber()
  if http.GET(module.RootURL) then
    local x = TXQuery.Create(http.Document)
    if module.website == 'HentaiFox' then
      page = tonumber(x.xpathstring('(//*[@class="pagination"]/li/a)[last()-1]'))
      if page == nil then page = 1 end
    else
      page = tonumber(x.xpathstring('//*[@class="pagination"]/a[last()-1]'))
      if page == nil then page = 1 end
    end
    return no_error
  else
    return net_problem
  end
end

function getimageurl()
  local u = MaybeFillHost(module.RootURL, task.PageContainerLinks[workid])
  if http.Get(u) then
    task.PageLinks[workid] = TXQuery.Create(http.document).XPathString('//div[@class="full_image"]//img/@src')
    return true
  end
  return false
end

function getnameandlink()
  if http.get(module.rooturl .. '/pag/' .. IncStr(url) .. '/') then
    local x = TXQuery.Create(http.Document)
    if module.website == 'HentaiFox' then
      x.xpathhrefall('//*[@class="lc_galleries"]//*[@class="caption"]//a', links, names)
    else
      x.xpathhrefall('//*[@class="preview_item"]/*[@class="caption"]/a', links, names)
    end
    return no_error
  else
    return net_problem
  end
end

function AddWebsiteModule(name, url)
  local m = NewModule()
  m.website = name
  m.rooturl = url
  m.category = 'H-Sites'
  m.sortedlist = true
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  m.ongetdirectorypagenumber = 'getdirectorypagenumber'
  m.ongetimageurl = 'getimageurl'
  return m
end

function Init()
  AddWebsiteModule('HentaiFox', 'https://hentaifox.com')
  AddWebsiteModule('AsmHentai', 'https://asmhentai.com')
end
